matter_server/server/sdk.py

```diff
@@ -168,7 +189,7 @@ async def open_commissioning_window(
     ) -> CommissioningParameters:
         """Open a commissioning window to commission a device present on this controller to another."""
         async with self._get_node_lock(node_id):
-            return await self._call_sdk(
+            return await self._call_sdk_non_reentrant(
```

Comment:
I was wondering if these are all cases where we need to ensure the same thread accesses the underlying sdk handle. If we are not entirely sure that the other cases are safe, why not just play it safe and always use this single-thread/instance executor all the time ? We have moved a lot of calls already to their async equivalents anyways already.

Reply:
Well, that is what I did in #599, but you've reverted it in #621, on the grounds of:

> Now that we have limited the executor to a single executor, we're blocking concurrent actions.
For example if we're busy resolving a node, we cant execute any device commands at the same time which gives a bad UX.

In the end it doesn't matter much, as there are only a few, afaik very short running calls left (`SetWiFiCredentials()`, `SetThreadOperationalDataset()`, `GetAddressAndPort()`, and some `Shutdown()` calls).

If we move all calls to the SDK to a single threaded executor, these can be (IMHO unnecessarily) be blocked by an ongoing Commissioning attempt.

Even when using the single-threaded executor for all SDK calls, we still call from two different Threads: the main asyncio event loop and the executor. Both use `Call()` underneath, which, if not re-entrant, could lead to the very same problem still :man_shrugging: 

In the end, this is all just a stop-gap solution: I want to convert all API's we use to proper co-routines. With that we will make call calls to the SDK from one Thread only (the asyncio event loop) avoid unnecessary Thread switches and side-step any re-entrancy issues.

Reply:
In the end, either way works for me. But I think this is the change with the least amount of possible surprises, while still fixing all re-entrance issues I am aware of. :sweat_smile:

Reply:
Yeah, I mean for all ongoing operations we already switched to the async variants already right ? I was just thinking that if we are not 100% certain about thread safety, we should play it safe. Also, not only threads but asyncio tasks can access objects from multiple tasks at once so it remains a tricky one